### ISSUE TITLE
front end caching

### ISSUE THREAD

**Issue by nickdesaulniers** [ClangBuiltLinux/continuous-integration2#308]
- [ ] can we propagate info between builds?
  - [ ] propagate HEAD shas for the kernel tree
  - [ ] propagate HEAD shas or version in for LLVM (maybe hardcode for released versions)
  - [ ] propagate prior run build status (red/green)
- [ ] at the start of a workflow, refetch SHAs
- [ ] if no change to SHAs, set exit status to prior build status
- [ ] rewrite latest SHAs for propagation

**Comment by nathanchance** [comment]
> can we propagate info between builds?

I think so. There are plenty of actions that will automatically commit to the repository:

https://github.com/devops-infra/action-commit-push
https://github.com/stefanzweifel/git-auto-commit-action
https://github.com/EndBug/add-and-commit

I do not see a way to do this without committing to the repository, we cannot access artifacts across runs (["Note: You can only download artifacts in a workflow that were uploaded during the same workflow run."](https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts#downloading-artifacts-during-a-workflow-run)) and I do not think we can write environment variables (at least, I have not found a way to in their documentation).

We could potentially write JSON or YAML with the compiler version, kernel version, and build status from the previous run then parse that when a new run starts.

Some initial thoughts:

* Will we run into issues where multiple workflows attempt to update their status files at the same time (`git push` will fail)? We could potentially run `git pull -r` before pushing? It looks like the last action I linked above has support for this.
* To get the current compiler version, we will need to use the TuxMake container, like we do in [`clang-version.yml`](https://github.com/ClangBuiltLinux/continuous-integration2/blob/15409e8d48ebe49e6908f406272a51c0988c2acc/.github/workflows/clang-version.yml).
* `git ls-remote <REPO> <BRANCH> | awk '{print $1}'` to get latest SHA without cloning:
  ```
  $ git ls-remote https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git linux-4.9.y | awk '{print $1}'
  9279031d74f8fe8760ce32ac527bc4658b578926
  ```

**Comment by nathanchance** [comment]
Talking with @kees about this during our weekly meeting, we think this is doable by giving every workflow file its own branch and information file, which eliminates the concern I had over multiple workflows updating one branch.

Rough idea for a `do-we-need-to-run.py` (or whatever we want to do):
- Does the workflow's branch exist in the repository?
  - If yes, clone the repo, checkout the branch, and read the file.
  - If no, create it, generate the information file, and exit with "we do need to build"
- Generate the new information file, including the current:
  - Kernel SHA
  - Compiler version string
- Has either item changed since the last run?
  - If yes, update the file, commit it, push it, and exit "we need to build"
  - If no, exit "we do not need to build"

I am not sure propagating the build status is necessary, since that flow does not seem to care about the result of the last build. I guess we would potentially want to retry builds that failed the previous time they were run in case things were flakey but it is possible we should just be re-running workflows manually in those cases and ignoring this check, which would simplify the overall implementation because we do not have to wait for the build results to update the file.

**Comment by nathanchance** [comment]
Initial take based on my comment above: https://github.com/ClangBuiltLinux/continuous-integration2/pull/522

**Comment by JustinStitt** [comment]
Here's a one month data dump in regards to potential caching opportunities:

This was from October 8th to November 8th 2023.

tl;dr: 17.1% of builds may have been skippable (assuming Kernel SHA and full compiler version are the only things that affect a build's outcome per job).

https://docs.google.com/spreadsheets/d/1Ag_N3kXYrBrBAq1VuXJ6ZAbBY8tHnWOaGV81VwaoGwk/edit?usp=sharing&resourcekey=0-OWVX-KZdJx23OOGn4nC-Cg

**Comment by nickdesaulniers** [comment]
Wow, thanks for running those statistics!  Some insights I get from the data (sort the sheet by column F):
- we're way overbuilding arm64-fixes.  That branch is not changing that often, so we would have had ~95% cache hits (if we had caching).  Until we do, I think it makes immediate sense to turn down the frequency of arm64-fixes builds.
- 5.10 (at least for Android and CrOS) and newer are changing pretty rapidly; the opportunities for caching is low.
- android-4.14 and android12-5.4 should be built less frequently.

No matter what, caching would be a win.

> So hey, that script is pretty good 😄

👍

**Comment by nathanchance** [comment]
> * we're way overbuilding arm64-fixes.  That branch is not changing that often, so we would have had ~95% cache hits (if we had caching).  Until we do, I think it makes immediate sense to turn down the frequency of arm64-fixes builds.

To be honest, it probably makes sense to just rely on -next testing for this one and just delete builds for it outright in my opinion. Doing the build once a week makes little sense to me. It gets [included in the fixes side of -next](https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/tree/Next/Trees) and the likelihood of a change getting added there that breaks LLVM and fast tracked to Linus without us noticing from -next is fairly low to me, as the arm64 folks are usually pretty good about holding fixes until they are in -next for a bit.

> * android-4.14 and android12-5.4 should be built less frequently.

4.14 goes EOL in January, so much agreed. We could reduce the LLVM ToT/stable builds of these trees to once a week, which would significantly reduce the number of builds of those trees.

> No matter what, caching would be a win.

Agreed. Maybe we sit down this weekend and think through implementation details?

**Comment by nathanchance** [comment]
Implemented in #664 :)
